### PR TITLE
Prevent the client from closing grpc connection by mistake

### LINKED ISSUE #11917: [Enhancement]: Prevent the client from closing grcp by mistake.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: lastest
- Deployment mode(standalone or cluster):
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

In the following code:
```
func (c *Client) recall(caller func() (interface{}, error)) (interface{}, error) {
	ret, err := caller()
	if err == nil {
		return ret, nil
	}
	log.Debug("IndexCoord Client grpc error", zap.Error(err))

	c.resetConnection()

	ret, err = caller()
	if err == nil {
		return ret, nil
	}
	return ret, err
}
```

When the caller returns an error for the first time, the current grpc connection will be forcibly closed.
In addition to the failure of the grpc link itself, the reason why the caller returns an error may also be the failure of ctx.
The failure of ctx may be due to cancellation or timeout.

Therefore, in the caller, it should be judged in advance whether ctx is invalid.
At the same time, it is determined in the recall function that when the returned err is equal to the ctx failure error, the error is directly returned instead of resetting the connection.

Taking IndexCoord's client as an example, the logic should be as follows:

```
func (c *Client) recall(caller func() (interface{}, error)) (interface{}, error) {
	ret, err := caller()
	if err == nil {
		return ret, nil
	}
	if err == context.Canceled || err == context.DeadlineExceeded {
		return nil, err
	}

	log.Debug("IndexCoord Client grpc error", zap.Error(err))

	c.resetConnection()

	ret, err = caller()
	if err == nil {
		return ret, nil
	}
	return ret, err
}


// BuildIndex sends the build index request to IndexCoord.
func (c *Client) BuildIndex(ctx context.Context, req *indexpb.BuildIndexRequest) (*indexpb.BuildIndexResponse, error) {
	ret, err := c.recall(func() (interface{}, error) {
		client, err := c.getGrpcClient()
		if err != nil {
			return nil, err
		}
		if !funcutil.CheckCtxValid(ctx) {
			return nil, ctx.Err()
		}
		return client.BuildIndex(ctx, req)
	})
	if err != nil || ret == nil {
		return nil, err
	}
	return ret.(*indexpb.BuildIndexResponse), err
}
```

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by czs007**
Resolves: #11917 
Related: #8521 #8824 
Signed-off-by: zhenshan.cao <zhenshan.cao@zilliz.com>

### CODE REVIEW

File: internal/distributed/datacoord/client/client.go [hunk 1]
```diff
@@ -458,16 +482,21 @@ func (c *Client) GetSegmentInfo(ctx context.Context, req *datapb.GetSegmentInfoR
...
-	client, err := c.getGrpcClient()
-	if err != nil {
-		return &commonpb.Status{
-			ErrorCode: commonpb.ErrorCode_UnexpectedError,
-			Reason:    err.Error(),
-		}, nil
-	}
 
-	return client.SaveBinlogPaths(ctx, req)
+	ret, err := c.recall(func() (interface{}, error) {
```
Comment (DragonDriver) [inline internal/distributed/datacoord/client/client.go:486]: I wonder why we don't recall `SaveBinlogPaths` before, but this patch does.

Comment (czs007) [reply internal/distributed/datacoord/client/client.go:486]: @XuanYang-cn  please check this.

File: internal/util/funcutil/func.go [hunk 2]
```diff
@@ -184,3 +184,7 @@ func GetAttrByKeyFromRepeatedKV(key string, kvs []*commonpb.KeyValuePair) (strin
 
 	return "", errors.New("key " + key + " not found")
 }
+
+func CheckCtxValid(ctx context.Context) bool {
```
Comment (DragonDriver) [inline internal/util/funcutil/func.go:188]: pls add an ut for `CheckCtxValid`.

Comment (czs007) [reply internal/util/funcutil/func.go:188]: @DragonDriver check this.

File: internal/distributed/datacoord/client/client.go (hunk 1 above)
Comment (XuanYang-cn) [reply internal/distributed/datacoord/client/client.go:486]: We don't need to recall this, if failed, DN will redo the flush.
Also we need to guarantee the order, if retry this, it'll break the order.

Comment (czs007) [reply internal/distributed/datacoord/client/client.go]: You won! @XuanYang-cn

**Comment by XuanYang-cn** [comment]
lgtm by me

**Pull request merged**